cms-contentful/src/main/java/com/commercetools/sunrise/cms/contentful/ContentfulCmsPage.java

```diff
@@ -1,92 +1,182 @@
+     * @return path segments of entries only
+     */
+    private String[] createEntryPathSegments(final String[] pathSegments) {
+        return copyOfRange(pathSegments, 0, pathSegments.length - 1);
     }
 
-    private Optional<String> getContent(final CDAEntry entry, final String contentFieldName) {
-        return getCdaField(entry, contentFieldName)
-                .flatMap(cdaField -> {
-                    final Object entryField = entry.getField(contentFieldName);
-                    return getContentAccordingToFieldDefinition(entryField, cdaField);
-                        });
+    /**
+     * Traverse contained CDAEntry to match all path segments.
+     *
+     * @param pathSegments array of all path segments that lead to CDAEntry of interest
+     * @return CDAEntry that matches path segments
+     */
+    private Optional<CDAEntry> findEntry(final String[] pathSegments) {
+        CDAEntry entry = cdaEntry;
+
+        for (String key: pathSegments) {
+            Object nextEntry = null;
+
+            Matcher arrayMatcher = ARRAY_KEY_PATTERN.matcher(key);
+
+            if (arrayMatcher.find()) {
+                nextEntry = getEntryFromArray(entry, arrayMatcher);
+            } else if (entry.rawFields().containsKey(key)) {
+                nextEntry = entry.getField(key);
+            }
+
+            if (nextEntry != null && nextEntry instanceof CDAEntry) {
+                entry = (CDAEntry) nextEntry;
+            } else {
+                return Optional.empty();
+            }
+        }
+
+        return Optional.of(entry);
     }
 
-    private CDAEntry getEntryWithContentField(@Nullable final CDAEntry cdaEntry, final List<String> entryNamesList) {
-        if (entryNamesList.isEmpty()) {
-            return cdaEntry;
-        } else {
-            final String key = entryNamesList.get(0);
-            if (cdaEntry != null && cdaEntry.rawFields().containsKey(key)) {
-                Object item = cdaEntry.getField(key);
-                if (item instanceof CDAEntry) {
-                    entryNamesList.remove(0);
-                    return getEntryWithContentField((CDAEntry) item, entryNamesList);
+    /**
+     * Try to get an item from input entry which is supposed to contain an array of fields.
+     *
+     * After pattern has been matched arrayMatcher contains a regexp group of key and index.
+     * E.g. after matching 'key[1]' two groups of 'key' and '1' are contained in matcher.
+     *
+     * They are later used to find 'key' field in parentEntry and retrieve '1' (second)
+     * item from that field which is expected to form an array list.
+     *
+     * If the process fails null is returned.
+     *
+     * @param parentEntry should contain expected array field
+     * @param arrayMatcher contains key and index groups after pattern has been matched
+     * @return matched entry or null
+     */
+    private Object getEntryFromArray(final CDAEntry parentEntry, final Matcher arrayMatcher) {
```

Comment:
Could you annotate the method with `@Nullable`? To minimize the possibilities of NPE.